File: content-solutions/dijkstra/route-path.java
import java.util.*;

class Solution {
    public int[] solution(int n, int[][] roads, int s, int e) {
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        for (int[] r : roads) {
            graph.get(r[0]).add(new int[] {r[1], r[2]});
            graph.get(r[1]).add(new int[] {r[0], r[2]});
        }
        long[] dist = new long[n];
        int[] prev = new int[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        Arrays.fill(prev, -1);
        dist[s] = 0;
        PriorityQueue<long[]> heap = new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));
        heap.offer(new long[] {0, s});
        while (!heap.isEmpty()) {
            long[] top = heap.poll();
            long d = top[0];
            int v = (int) top[1];
            if (d > dist[v]) continue;
            for (int[] x : graph.get(v)) {
                if (d + x[1] < dist[x[0]]) {
                    dist[x[0]] = d + x[1];
                    prev[x[0]] = v;
                    heap.offer(new long[] {dist[x[0]], x[0]});
                }
            }
        }
        if (dist[e] == Long.MAX_VALUE) return new int[0];
        List<Integer> path = new ArrayList<>();
        for (int v = e; v != -1; v = prev[v]) path.add(v);
        Collections.reverse(path);
        int[] answer = new int[path.size()];
        for (int i = 0; i < answer.length; i++) answer[i] = path.get(i);
        return answer;
    }
}

File: content-solutions/dijkstra/route-path.py
import heapq


def solution(n, roads, s, e):
    graph = [[] for _ in range(n)]
    for a, b, t in roads:
        graph[a].append((b, t))
        graph[b].append((a, t))
    dist = [float("inf")] * n
    prev = [-1] * n
    dist[s] = 0
    heap = [(0, s)]
    while heap:
        d, v = heapq.heappop(heap)
        if d > dist[v]:
            continue
        for w, t in graph[v]:
            if d + t < dist[w]:
                dist[w] = d + t
                prev[w] = v
                heapq.heappush(heap, (dist[w], w))
    if dist[e] == float("inf"):
        return []
    path = []
    v = e
    while v != -1:
        path.append(v)
        v = prev[v]
    return path[::-1]
